File: AssistantPlus/Application/Utility/MiscellaneousHelper.cs
#pragma warning disable 0,

using Twinning.AssistantPlus;

namespace Twinning.AssistantPlus.Utility {

	public static class MiscellaneousHelper {

		#region storage

		public static async Task<List<StoragePath>> PickStorageItem(
			String          tag,
			StoragePickType type,
			Boolean         multiply,
			StoragePath?    location,
			String?         name
		) {
			if (location == null) {
				location = App.Instance.Setting.Data.StoragePickerHistoryLocation.GetValueOrDefault(tag);
			}
			var target = await StorageHelper.Pick(type, multiply, location, name, App.Instance.MainWindow);
			if (!target.IsEmpty()) {
				App.Instance.Setting.Data.StoragePickerHistoryLocation[tag] = type switch {
					StoragePickType.LoadFile      => target.First().Parent().AsNotNull(),
					StoragePickType.LoadDirectory => target.First(),
					StoragePickType.SaveFile      => target.First().Parent().AsNotNull(),
					_                             => throw new UnreachableException(),
				};
				await App.Instance.Setting.Save(apply: false);
			}
			return target;
		}

		#endregion

	}

}

File: Assistant/lib/utility/miscellaneous_helper.dart
import '/common.dart';
import '/setting.dart';
import '/utility/storage_path.dart';
import '/utility/storage_helper.dart';
import '/widget/control.dart';
import 'package:flutter/widgets.dart';
import 'package:provider/provider.dart';

// ----------------

class MiscellaneousHelper {

  // #region storage

  static Future<List<StoragePath>> pickStorageItem(
    BuildContext          context,
    String                tag,
    List<StoragePickType> type,
    Boolean               multiply,
    StoragePath?          location,
    String?               name, {
    TextStyle?            textStyle = null, // TODO: remove?
  }) async {
    var setting = Provider.of<SettingProvider>(context, listen: false);
    var actualType = null as StoragePickType?;
    assertTest(!type.isEmpty);
    if (type.length == 1) {
      actualType = type.first;
    }
    else {
      actualType = await StyledMenuExtension.show<StoragePickType>(context, StyledMenu.standard(
        position: .under,
        content: <({StoragePickType value, String text})>[
          if (type.contains(StoragePickType.loadFile)) ...[
            (value: .loadFile, text: 'Load File'),
          ],
          if (type.contains(StoragePickType.loadDirectory)) ...[
            (value: .loadDirectory, text: 'Load Directory'),
          ],
          if (type.contains(StoragePickType.saveFile)) ...[
            (value: .saveFile, text: 'Save File'),
          ],
        ].map((value) => StyledMenuItem.standard(
          value: value.value,
          content: StyledText.custom(
            value.text,
            style: textStyle,
          ),
        )),
      ));
    }
    if (location == null) {
      location = setting.data.storagePickerHistoryLocation[tag];
    }
    var target = actualType == null ? <StoragePath>[] : await StorageHelper.pick(actualType, multiply, location, name);
    if (!target.isEmpty) {
      setting.data.storagePickerHistoryLocation[tag] = switch (actualType!) {
        .loadFile      => target.first.parent()!,
        .loadDirectory => target.first,
        .saveFile      => target.first.parent()!,
      };
      await setting.save(apply: false);
    }
    return target;
  }

  // #endregion

}
